<a href="https://colab.research.google.com/github/Koks-creator/DamagedRoadSystem/blob/main/yolov11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')
!ln -s /content/gdrive/My\ Drive/ /mydrive

Mounted at /content/gdrive


In [ ]:
!unzip /mydrive/yoloTrain/train_data.zip -d /content

Archive:  /mydrive/yoloTrain/train_data.zip
   creating: /content/train_data/
   creating: /content/train_data/images/
   creating: /content/train_data/images/train/
  inflating: /content/train_data/images/train/nwm2_0_0.jpg  
  inflating: /content/train_data/images/train/nwm2_0_11.png  
  inflating: /content/train_data/images/train/nwm2_0_15.jpg  
  inflating: /content/train_data/images/train/nwm2_0_16.jpg  
  inflating: /content/train_data/images/train/nwm2_0_17.jpg  
  inflating: /content/train_data/images/train/nwm2_0_18.jpg  
  inflating: /content/train_data/images/train/nwm2_0_19.jpeg  
  inflating: /content/train_data/images/train/nwm2_0_21.jpg  
  inflating: /content/train_data/images/train/nwm2_0_22.jpg  
  inflating: /content/train_data/images/train/nwm2_0_24.jpg  
  inflating: /content/train_data/images/train/nwm2_0_26.jpg  
  inflating: /content/train_data/images/train/nwm2_0_27.jpeg  
  inflating: /content/train_data/images/train/nwm2_0_28.jpg  
  inflating: /content/train

In [ ]:
!pip install ultralytics==8.3.89 sahi==0.11.22

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.2/87.2 kB 5.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 932.6/932.6 kB 33.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.2/114.2 kB 12.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.5/62.5 MB 15.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 92.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 64.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 55.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 2.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 14.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 8.5 MB/s eta 0:00:00


In [ ]:
!pip freeze | grep -e ultralytics -e sahi

sahi==0.11.22
ultralytics==8.3.89
ultralytics-thop==2.0.14


In [ ]:
from ultralytics import YOLO
from sahi.utils.ultralytics import download_yolo11n_model
import yaml
import os
import torch
import shutil

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.


In [ ]:
if torch.cuda.is_available():
    device_count = torch.cuda.device_count()
    device_name = torch.cuda.get_device_name(0) if device_count > 0 else "None"
    print(f"Found {device_count} GPU devices.")
    print(f"Active device: {device_name}")
    device = 0  # Użyj pierwszego GPU
else:
    print("GPU not found, using CPU.")
    device = 'cpu'

Found 1 GPU devices.
Active device: Tesla T4


In [ ]:
DATASET_PATH = "/content/train_data"
CLASS_NAMES = ["hole", "crack", "block cracking", "patching"]
MODEL_DIR = "./models"
MODEL_NAME = "yolov11m.pt"
MODEL_PATH = os.path.join(MODEL_DIR, MODEL_NAME)
BATCH_SIZE = 16
WORKERS = 8
CONFIG_YAML_FILE = "./config/data.yaml"

In [ ]:
os.makedirs(MODEL_DIR, exist_ok=True)

if not os.path.exists(MODEL_PATH):
    print(f"Model {MODEL_NAME} not found. Downloading...")

    download_yolo11n_model(MODEL_PATH)
    print(f"Model has been downloaded, to {MODEL_PATH}")

Model yolov11m.pt not found. Downloading...


yolov11m.pt: 100%|██████████| 5.35M/5.35M [00:00<00:00, 106MB/s]

Model has been downloaded, to ./models/yolov11m.pt


In [ ]:
data_yaml = {
    'path': DATASET_PATH,
    'train': 'images/train',
    'val': 'images/val',
    'names': {i: name for i, name in enumerate(CLASS_NAMES)},
    'nc': len(CLASS_NAMES)
}

os.makedirs("config", exist_ok=True)
with open(CONFIG_YAML_FILE, 'w') as f:
    yaml.dump(data_yaml, f, sort_keys=False)

print(f"Config file created: {CONFIG_YAML_FILE}")

Config file created: ./config/data.yaml


In [ ]:
print(f"Loading model from {MODEL_PATH}...")
model = YOLO(MODEL_PATH)

Loading model from ./models/yolov11m.pt...


In [ ]:
results = model.train(
    data=CONFIG_YAML_FILE,
    epochs=80,
    imgsz=800,
    batch=BATCH_SIZE,
    patience=30,
    project="runs",
    name="yolov11_custom",
    pretrained=True,
    optimizer='SGD',
    momentum=0.937,
    weight_decay=0.0005,
    lr0=0.0015,       # More moderate learning rate - balance between stability and recall improvement
    lrf=0.005,        # Gentler LR decay - less aggressive
    warmup_epochs=5,  # Shorter warmup - faster transition to full LR
    cos_lr=True,
    augment=True,
    mosaic=0.7,       # Less intense mosaic - cracks may disappear with too strong mosaic
    mixup=0.15,       # Less intense mixup - preserves more distinct crack features
    copy_paste=0.2,   # Remains high - good for rare classes
    degrees=8.0,      # Reduced rotation - cracks may have directional characteristics
    scale=0.6,        # Larger scale - various crack sizes
    translate=0.2,    # Added translation - helps with different crack locations
    fliplr=0.5,       # Added horizontal flip - increases data diversity
    hsv_h=0.015,
    hsv_s=0.7,
    hsv_v=0.4,
    perspective=0.0005,
    shear=1.0,        # Reduced shearing - preserves natural appearance of cracks
    cls=1.0,          # Higher weight for classification - crucial for recall
    box=3.5,          # Lower weight for box loss - less emphasis on location accuracy
    iou=0.35,         # Lower IoU threshold - increases recall at the cost of precision
    save=True,
    save_period=5,    # More frequent model saving
    device=device,
    workers=WORKERS
)

# results = model.train(
#     data=CONFIG_YAML_FILE,
#     epochs=80,
#     imgsz=800,
#     batch=BATCH_SIZE,
#     patience=30,
#     project="runs",
#     name="yolov11_custom",
#     pretrained=True,
#     optimizer='SGD',
#     momentum=0.937,
#     weight_decay=0.0005,
#     lr0=0.0015,       # More moderate learning rate - balance between stability and recall improvement
#     lrf=0.005,        # Gentler LR decay - less aggressive
#     warmup_epochs=5,  # Shorter warmup - faster transition to full LR
#     cos_lr=True,
#     augment=True,
#     mosaic=0.7,       # Less intense mosaic - cracks may disappear with too strong mosaic
#     mixup=0.15,       # Less intense mixup - preserves more distinct crack features
#     copy_paste=0.2,   # Remains high - good for rare classes
#     degrees=8.0,      # Reduced rotation - cracks may have directional characteristics
#     scale=0.6,        # Larger scale - various crack sizes
#     translate=0.2,    # Added translation - helps with different crack locations
#     fliplr=0.5,       # Added horizontal flip - increases data diversity
#     hsv_h=0.015,
#     hsv_s=0.7,
#     hsv_v=0.4,
#     perspective=0.0005,
#     shear=1.0,        # Reduced shearing - preserves natural appearance of cracks
#     cls=1.0,          # Higher weight for classification - crucial for recall
#     box=3.5,          # Lower weight for box loss - less emphasis on location accuracy
#     iou=0.35,         # Lower IoU threshold - increases recall at the cost of precision
#     save=True,
#     save_period=5,    # More frequent model saving
#     device=device,
#     workers=WORKERS
# )

New https://pypi.org/project/ultralytics/8.3.130 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.89 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=./models/yolov11m.pt, data=./config/data.yaml, epochs=80, time=None, patience=30, batch=16, imgsz=800, save=True, save_period=5, cache=False, device=0, workers=8, project=runs, name=yolov11_custom, exist_ok=False, pretrained=True, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=True, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.35, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=True, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_fr

100%|██████████| 755k/755k [00:00<00:00, 20.9MB/s]


Overriding model.yaml nc=80 with nc=4

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      6640  ultralytics.nn.modules.block.C3k2            [32, 64, 1, False, 0.25]      
  3                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
  4                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  5                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  6                  -1  1     87040  ultralytics.nn.modules.block.C3k2            [128, 128, 1, True]           
  7                  -1  1    295424  ultralytics

100%|██████████| 5.35M/5.35M [00:00<00:00, 90.9MB/s]


AMP: checks passed ✅


train: Scanning /content/train_data/labels/train... 790 images, 0 backgrounds, 0 corrupt: 100%|██████████| 790/790 [00:01<00:00, 433.76it/s]

train: WARNING ⚠️ /content/train_data/images/train/nwm_2_10.jpg: corrupt JPEG restored and saved
train: New cache created: /content/train_data/labels/train.cache


albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))


val: Scanning /content/train_data/labels/val... 26 images, 0 backgrounds, 0 corrupt: 100%|██████████| 26/26 [00:00<00:00, 231.74it/s]

val: New cache created: /content/train_data/labels/val.cache


Plotting labels to runs/yolov11_custom/labels.jpg... 
optimizer: SGD(lr=0.0015, momentum=0.937) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 800 train, 800 val
Using 2 dataloader workers
Logging results to runs/yolov11_custom
Starting training for 80 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/80      3.83G     0.9894      8.072      2.042         43        800: 100%|██████████| 50/50 [00:31<00:00,  1.57it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:02<00:00,  2.10s/it]

                   all         26         80    0.00481      0.467    0.00961    0.00294



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/80      4.04G     0.9181      7.733       1.92         18        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.43it/s]

                   all         26         80    0.00486      0.389     0.0113    0.00416



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/80      4.05G     0.8908      7.168      1.865         42        800: 100%|██████████| 50/50 [00:29<00:00,  1.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.42it/s]

                   all         26         80      0.353      0.133     0.0297    0.00895



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/80      4.06G     0.8695      6.749      1.821         55        800: 100%|██████████| 50/50 [00:29<00:00,  1.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.56it/s]

                   all         26         80      0.934      0.133      0.172      0.102



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/80      4.07G     0.8683       6.28      1.791         35        800: 100%|██████████| 50/50 [00:28<00:00,  1.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.57it/s]

                   all         26         80      0.903        0.2      0.242      0.113



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/80      4.07G     0.8761      5.981      1.795         40        800: 100%|██████████| 50/50 [00:29<00:00,  1.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.14it/s]

                   all         26         80      0.452      0.264      0.271      0.138



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/80      4.09G     0.8429      5.643      1.722         25        800: 100%|██████████| 50/50 [00:28<00:00,  1.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.26it/s]

                   all         26         80      0.595      0.361      0.371      0.175



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/80      4.09G     0.8378      5.314      1.683         32        800: 100%|██████████| 50/50 [00:28<00:00,  1.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.83it/s]

                   all         26         80      0.417      0.359      0.289      0.145



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/80      4.11G     0.8406      5.251      1.712         38        800: 100%|██████████| 50/50 [00:28<00:00,  1.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.57it/s]

                   all         26         80      0.522      0.402      0.365       0.13



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/80      4.11G     0.8054      4.959       1.64         33        800: 100%|██████████| 50/50 [00:27<00:00,  1.83it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.50it/s]

                   all         26         80      0.434      0.376      0.359      0.183



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/80      4.13G      0.805      4.774      1.627         31        800: 100%|██████████| 50/50 [00:29<00:00,  1.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.54it/s]

                   all         26         80      0.608      0.356      0.432      0.176



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/80      4.13G     0.8034      4.812      1.639         28        800: 100%|██████████| 50/50 [00:27<00:00,  1.81it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.52it/s]

                   all         26         80      0.331      0.344      0.308      0.125



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/80      4.15G     0.8147      4.699      1.633         55        800: 100%|██████████| 50/50 [00:26<00:00,  1.85it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.72it/s]

                   all         26         80      0.452       0.43       0.45      0.177



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/80      4.15G     0.7911      4.516      1.593         29        800: 100%|██████████| 50/50 [00:28<00:00,  1.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.60it/s]

                   all         26         80      0.509      0.368      0.369      0.169



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/80      4.17G     0.7797      4.471      1.603         45        800: 100%|██████████| 50/50 [00:27<00:00,  1.81it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.61it/s]

                   all         26         80      0.506      0.322      0.337       0.14



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/80      4.17G     0.7982      4.435      1.607         59        800: 100%|██████████| 50/50 [00:28<00:00,  1.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.66it/s]

                   all         26         80      0.607      0.389      0.425      0.201



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/80      4.19G     0.7871      4.416       1.59         79        800: 100%|██████████| 50/50 [00:28<00:00,  1.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.94it/s]

                   all         26         80        0.5      0.383      0.334      0.154



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/80      4.19G     0.7889      4.372      1.609         30        800: 100%|██████████| 50/50 [00:30<00:00,  1.66it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.59it/s]

                   all         26         80      0.333      0.433      0.349      0.141



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/80      4.21G     0.7852       4.32      1.585         44        800: 100%|██████████| 50/50 [00:28<00:00,  1.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.43it/s]

                   all         26         80      0.481      0.428      0.437      0.194



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/80      4.21G     0.7805      4.327      1.585         24        800: 100%|██████████| 50/50 [00:29<00:00,  1.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.46it/s]

                   all         26         80      0.608      0.449      0.441      0.157



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/80      4.23G      0.764      4.216      1.577         15        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.36it/s]

                   all         26         80      0.516      0.371       0.45      0.187



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/80      4.23G     0.7705      4.109      1.545         56        800: 100%|██████████| 50/50 [00:28<00:00,  1.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.55it/s]

                   all         26         80      0.562       0.37      0.432      0.205



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/80      4.25G     0.7711      4.195       1.57         36        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.58it/s]

                   all         26         80       0.69      0.472      0.465      0.238



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/80      4.25G     0.7643       4.11      1.569         27        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.54it/s]

                   all         26         80      0.723      0.441      0.488      0.227



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/80      4.27G     0.7689      4.141      1.572         55        800: 100%|██████████| 50/50 [00:30<00:00,  1.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.54it/s]

                   all         26         80      0.489      0.339      0.364       0.19



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/80      4.27G     0.7798      4.127      1.582         48        800: 100%|██████████| 50/50 [00:28<00:00,  1.72it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.15it/s]

                   all         26         80      0.511      0.393      0.409      0.206



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/80      4.29G     0.7566      3.941      1.537         24        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.61it/s]

                   all         26         80      0.524      0.378      0.452      0.188



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/80      4.29G     0.7607      3.953      1.547         52        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.31it/s]

                   all         26         80      0.495      0.497        0.5      0.212



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/80      4.31G     0.7333      3.865      1.517         52        800: 100%|██████████| 50/50 [00:28<00:00,  1.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.81it/s]

                   all         26         80      0.518      0.461       0.47       0.23



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/80      4.31G     0.7551      3.933       1.54         61        800: 100%|██████████| 50/50 [00:28<00:00,  1.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.43it/s]

                   all         26         80      0.657        0.4      0.462       0.24



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/80      4.33G     0.7615      3.932      1.544         37        800: 100%|██████████| 50/50 [00:30<00:00,  1.66it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.57it/s]

                   all         26         80      0.497       0.45      0.459      0.242



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/80      4.33G      0.735      3.811      1.512         18        800: 100%|██████████| 50/50 [00:30<00:00,  1.62it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.57it/s]

                   all         26         80      0.527      0.419      0.462      0.238



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/80      4.35G     0.7573      3.839      1.527         53        800: 100%|██████████| 50/50 [00:29<00:00,  1.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.46it/s]

                   all         26         80      0.602      0.383      0.447      0.211



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/80      4.35G     0.7563      3.824      1.532         41        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.27it/s]

                   all         26         80      0.518      0.428      0.424      0.232



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/80      4.37G     0.7507      3.814      1.543         70        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.46it/s]

                   all         26         80      0.595      0.423      0.461       0.22



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/80      4.37G      0.746      3.765      1.521         23        800: 100%|██████████| 50/50 [00:29<00:00,  1.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.51it/s]

                   all         26         80      0.569      0.361      0.413       0.21



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/80      4.39G     0.7495      3.741      1.538         51        800: 100%|██████████| 50/50 [00:28<00:00,  1.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.71it/s]

                   all         26         80      0.625      0.339      0.494      0.232



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/80      4.39G     0.7391      3.647      1.503         36        800: 100%|██████████| 50/50 [00:28<00:00,  1.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.79it/s]

                   all         26         80      0.653      0.361      0.455      0.227



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/80      4.41G     0.7359       3.65      1.513         61        800: 100%|██████████| 50/50 [00:30<00:00,  1.66it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.60it/s]

                   all         26         80      0.729      0.387      0.452      0.246



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/80      4.41G     0.7476      3.663      1.504         38        800: 100%|██████████| 50/50 [00:29<00:00,  1.72it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.90it/s]

                   all         26         80      0.642      0.421      0.459      0.223



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/80      4.43G     0.7341      3.682      1.513         16        800: 100%|██████████| 50/50 [00:28<00:00,  1.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.70it/s]

                   all         26         80      0.512      0.428      0.419       0.19



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/80      4.43G     0.7437      3.692      1.526         60        800: 100%|██████████| 50/50 [00:29<00:00,  1.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.54it/s]

                   all         26         80      0.537      0.498      0.462      0.213



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/80      4.45G     0.7299      3.542      1.483        104        800: 100%|██████████| 50/50 [00:28<00:00,  1.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.52it/s]

                   all         26         80       0.59      0.439      0.454      0.197



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/80      4.45G     0.7386      3.575      1.497         47        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.90it/s]

                   all         26         80      0.688      0.417      0.495      0.219



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/80      4.47G     0.7228      3.429      1.459         37        800: 100%|██████████| 50/50 [00:30<00:00,  1.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.36it/s]

                   all         26         80      0.718      0.437      0.521      0.221



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/80      4.47G     0.7294      3.554      1.494         62        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.65it/s]

                   all         26         80      0.757      0.424      0.521       0.23



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/80      4.49G     0.7226      3.442      1.457         42        800: 100%|██████████| 50/50 [00:28<00:00,  1.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.83it/s]

                   all         26         80      0.556      0.467      0.482      0.201



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/80      4.49G      0.716       3.48      1.479         60        800: 100%|██████████| 50/50 [00:27<00:00,  1.81it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.91it/s]

                   all         26         80      0.747      0.413      0.514      0.222



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/80      4.51G      0.716      3.505      1.483         27        800: 100%|██████████| 50/50 [00:28<00:00,  1.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.02it/s]

                   all         26         80      0.501      0.451      0.471      0.214



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/80      4.51G     0.7257      3.555      1.502         80        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.61it/s]

                   all         26         80      0.737      0.441      0.504      0.227



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      51/80      4.53G     0.7138      3.415      1.485         47        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.49it/s]

                   all         26         80       0.74      0.396      0.505      0.247



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      52/80      4.53G     0.7284      3.493      1.479         52        800: 100%|██████████| 50/50 [00:30<00:00,  1.66it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.73it/s]

                   all         26         80      0.529      0.548        0.5      0.213



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      53/80      4.54G     0.7255      3.436      1.482         24        800: 100%|██████████| 50/50 [00:28<00:00,  1.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.36it/s]

                   all         26         80      0.504      0.456       0.49      0.216



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      54/80      4.55G     0.7166      3.411      1.463         24        800: 100%|██████████| 50/50 [00:28<00:00,  1.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.45it/s]

                   all         26         80      0.548      0.428      0.478      0.218



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      55/80      4.56G     0.7211      3.345      1.455         27        800: 100%|██████████| 50/50 [00:28<00:00,  1.75it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.70it/s]

                   all         26         80      0.559      0.488       0.51      0.236



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      56/80      4.57G     0.7132      3.438      1.479         34        800: 100%|██████████| 50/50 [00:29<00:00,  1.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.16it/s]

                   all         26         80      0.638      0.431       0.51      0.238



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      57/80      4.59G     0.7036      3.262      1.429         48        800: 100%|██████████| 50/50 [00:27<00:00,  1.80it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.06it/s]

                   all         26         80      0.648       0.46      0.494      0.233



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      58/80      4.59G     0.7123      3.336      1.461         45        800: 100%|██████████| 50/50 [00:28<00:00,  1.73it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.69it/s]

                   all         26         80      0.657      0.453      0.508      0.235



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      59/80      4.61G     0.7056      3.334      1.442         33        800: 100%|██████████| 50/50 [00:30<00:00,  1.65it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.83it/s]

                   all         26         80      0.633      0.453      0.499      0.246



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      60/80      4.61G     0.7168      3.359      1.465         47        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.74it/s]

                   all         26         80      0.661      0.448      0.507       0.24



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      61/80      4.62G     0.7038      3.269      1.441         25        800: 100%|██████████| 50/50 [00:29<00:00,  1.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.56it/s]

                   all         26         80      0.623      0.428      0.486      0.241



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      62/80      4.63G     0.7137      3.371      1.456         55        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.73it/s]

                   all         26         80      0.702       0.42      0.498      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      63/80      4.64G     0.7061      3.279      1.444         47        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.81it/s]

                   all         26         80      0.669      0.446      0.493      0.216



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      64/80      4.65G     0.7014      3.268      1.444         43        800: 100%|██████████| 50/50 [00:29<00:00,  1.68it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.71it/s]

                   all         26         80       0.72      0.467      0.505      0.244



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      65/80      4.67G     0.7164      3.383      1.479         68        800: 100%|██████████| 50/50 [00:29<00:00,  1.72it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.32it/s]

                   all         26         80      0.652      0.439      0.486      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      66/80      4.67G     0.7074      3.316      1.459         29        800: 100%|██████████| 50/50 [00:29<00:00,  1.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.38it/s]

                   all         26         80      0.701      0.421      0.491      0.242



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      67/80      4.68G      0.691      3.204      1.431         38        800: 100%|██████████| 50/50 [00:28<00:00,  1.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.45it/s]

                   all         26         80      0.638      0.406      0.489      0.242



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      68/80      4.69G     0.7215      3.358       1.47         34        800: 100%|██████████| 50/50 [00:28<00:00,  1.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.86it/s]

                   all         26         80      0.716      0.432      0.511      0.248



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      69/80       4.7G     0.7079      3.248      1.445         60        800: 100%|██████████| 50/50 [00:29<00:00,  1.71it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.21it/s]

                   all         26         80      0.708      0.416      0.508      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      70/80      4.71G     0.7014      3.186      1.445         52        800: 100%|██████████| 50/50 [00:29<00:00,  1.70it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.55it/s]

                   all         26         80      0.727      0.393      0.502      0.242


Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      71/80      4.72G     0.6416      3.208      1.362         14        800: 100%|██████████| 50/50 [00:28<00:00,  1.76it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.23it/s]

                   all         26         80      0.444      0.544      0.508      0.243



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      72/80      4.73G     0.6471      3.169      1.361         10        800: 100%|██████████| 50/50 [00:26<00:00,  1.91it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.10it/s]

                   all         26         80       0.46      0.522      0.483      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      73/80      4.74G       0.64       3.08      1.344         17        800: 100%|██████████| 50/50 [00:26<00:00,  1.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.44it/s]

                   all         26         80      0.453      0.495      0.487      0.243



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      74/80      4.75G     0.6367       3.08      1.341          7        800: 100%|██████████| 50/50 [00:25<00:00,  1.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.60it/s]

                   all         26         80      0.454      0.481      0.463      0.239



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      75/80      4.76G     0.6368      3.107      1.339         58        800: 100%|██████████| 50/50 [00:25<00:00,  1.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.63it/s]

                   all         26         80      0.525      0.462      0.492      0.246



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      76/80      4.77G     0.6419      3.064      1.338         17        800: 100%|██████████| 50/50 [00:25<00:00,  1.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.54it/s]

                   all         26         80      0.502      0.472      0.489      0.243



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      77/80      4.79G     0.6404      3.031      1.336         12        800: 100%|██████████| 50/50 [00:25<00:00,  1.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.72it/s]

                   all         26         80      0.488      0.484      0.487      0.242



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      78/80      4.79G     0.6404      3.024      1.344         16        800: 100%|██████████| 50/50 [00:25<00:00,  1.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  2.02it/s]

                   all         26         80      0.564      0.479      0.497      0.247



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      79/80      4.81G     0.6376      3.066      1.342          7        800: 100%|██████████| 50/50 [00:25<00:00,  1.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.47it/s]

                   all         26         80      0.492      0.489      0.496      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      80/80      4.81G     0.6406      3.035      1.343         40        800: 100%|██████████| 50/50 [00:26<00:00,  1.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.58it/s]

                   all         26         80      0.521      0.489      0.495      0.246



80 epochs completed in 0.666 hours.
Optimizer stripped from runs/yolov11_custom/weights/last.pt, 5.5MB
Optimizer stripped from runs/yolov11_custom/weights/best.pt, 5.5MB

Validating runs/yolov11_custom/weights/best.pt...
Ultralytics 8.3.89 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
YOLO11n summary (fused): 100 layers, 2,582,932 parameters, 0 gradients, 6.3 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.88it/s]


                   all         26         80       0.87      0.333      0.474      0.229
                  hole          3          5      0.834        0.6      0.611      0.311
                 crack         23         60      0.819        0.2      0.448      0.194
        block cracking          9         15      0.955        0.2      0.364      0.182
Speed: 0.3ms preprocess, 14.9ms inference, 0.0ms loss, 1.8ms postprocess per image
Results saved to runs/yolov11_custom


In [ ]:
model = YOLO('/content/runs/yolov11_custom/weights/best.pt')  # Load the best model from previous training
results = model.train(
    data=CONFIG_YAML_FILE,
    epochs=50,
    imgsz=800,
    batch=max(4, BATCH_SIZE//2),  # Smaller batch size for better generalization
    patience=20,
    project="runs",
    name="yolov11_custom_finetuned",
    optimizer='SGD',
    momentum=0.9,
    weight_decay=0.0005,
    lr0=0.0005,        # Very low LR for fine-tuning
    lrf=0.001,
    cos_lr=True,
    augment=True,
    mosaic=0.1,        # Less intense mosaic
    mixup=0.4,
    copy_paste=0.6,    # Very intense copy-paste for rare classes
    cls=3,           # Very high weight for classification
    box=2.5,           # Low weight for box loss
    iou=0.3,           # Lower IoU threshold
    device=device,
    workers=WORKERS
)

New https://pypi.org/project/ultralytics/8.3.130 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.89 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: task=detect, mode=train, model=/content/runs/yolov11_custom/weights/best.pt, data=./config/data.yaml, epochs=50, time=None, patience=20, batch=8, imgsz=800, save=True, save_period=-1, cache=False, device=0, workers=8, project=runs, name=yolov11_custom_finetuned, exist_ok=False, pretrained=True, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=True, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.3, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=True, agnostic_nms=False, classes=None, retina_masks=False

train: Scanning /content/train_data/labels/train.cache... 790 images, 0 backgrounds, 0 corrupt: 100%|██████████| 790/790 [00:00<?, ?it/s]

train: WARNING ⚠️ /content/train_data/images/train/nwm_2_10.jpg: corrupt JPEG restored and saved
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))



val: Scanning /content/train_data/labels/val.cache... 26 images, 0 backgrounds, 0 corrupt: 100%|██████████| 26/26 [00:00<?, ?it/s]


Plotting labels to runs/yolov11_custom_finetuned/labels.jpg... 
optimizer: SGD(lr=0.0005, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 800 train, 800 val
Using 2 dataloader workers
Logging results to runs/yolov11_custom_finetuned
Starting training for 50 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/50      2.27G     0.6207      11.85       1.64         46        800: 100%|██████████| 99/99 [00:36<00:00,  2.74it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:01<00:00,  1.70it/s]

                   all         26         80      0.447      0.439      0.437      0.232



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/50      2.33G      0.556      11.03       1.54         46        800: 100%|██████████| 99/99 [00:33<00:00,  2.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.09it/s]

                   all         26         80       0.49      0.561      0.496      0.253



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/50      2.33G     0.5373      10.99      1.547         33        800: 100%|██████████| 99/99 [00:33<00:00,  2.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.04it/s]

                   all         26         80      0.517      0.532      0.457      0.252



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/50      2.33G     0.5279      11.03      1.531         38        800: 100%|██████████| 99/99 [00:33<00:00,  2.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.83it/s]

                   all         26         80      0.552      0.361      0.407      0.234



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/50       2.4G     0.5245      10.89       1.51         24        800: 100%|██████████| 99/99 [00:32<00:00,  3.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.58it/s]

                   all         26         80      0.444      0.433       0.42      0.248



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/50       2.4G     0.5242      11.09      1.541         15        800: 100%|██████████| 99/99 [00:34<00:00,  2.90it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.42it/s]

                   all         26         80      0.505      0.504      0.468      0.264



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/50       2.4G     0.4983      10.46      1.476         16        800: 100%|██████████| 99/99 [00:31<00:00,  3.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.24it/s]

                   all         26         80      0.519      0.517      0.473      0.256



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/50       2.4G     0.5149      10.81      1.519         36        800: 100%|██████████| 99/99 [00:32<00:00,  3.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.41it/s]

                   all         26         80      0.521      0.522      0.484      0.276



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/50       2.4G     0.5146      10.72      1.504         33        800: 100%|██████████| 99/99 [00:35<00:00,  2.82it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.19it/s]

                   all         26         80      0.533      0.546      0.504      0.279



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/50       2.4G     0.5119       10.7      1.488         33        800: 100%|██████████| 99/99 [00:31<00:00,  3.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.67it/s]

                   all         26         80      0.611      0.537      0.478       0.26



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/50       2.4G      0.504      10.38      1.465         20        800: 100%|██████████| 99/99 [00:33<00:00,  2.94it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.07it/s]

                   all         26         80      0.607      0.535      0.509       0.27



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/50       2.4G     0.5125       10.7      1.491         46        800: 100%|██████████| 99/99 [00:32<00:00,  3.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  5.26it/s]


                   all         26         80      0.644      0.434      0.467      0.253

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/50       2.4G     0.5093       10.7      1.488         31        800: 100%|██████████| 99/99 [00:32<00:00,  3.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  5.00it/s]

                   all         26         80       0.65       0.49      0.499      0.258



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/50       2.4G      0.517      10.78      1.515         26        800: 100%|██████████| 99/99 [00:34<00:00,  2.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.61it/s]

                   all         26         80      0.551      0.543      0.509      0.268



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/50       2.4G     0.5015      10.36      1.472         32        800: 100%|██████████| 99/99 [00:34<00:00,  2.88it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.72it/s]

                   all         26         80      0.524      0.511      0.472      0.265



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/50       2.4G     0.5192      10.78      1.523         32        800: 100%|██████████| 99/99 [00:33<00:00,  2.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  5.63it/s]

                   all         26         80      0.479        0.5      0.478      0.274



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/50       2.4G     0.5147      10.78      1.525         48        800: 100%|██████████| 99/99 [00:33<00:00,  3.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.86it/s]

                   all         26         80      0.574      0.516      0.496      0.272



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/50       2.4G     0.5036      10.44      1.495         36        800: 100%|██████████| 99/99 [00:33<00:00,  3.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.55it/s]

                   all         26         80      0.493      0.512      0.487      0.246



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/50       2.4G     0.5075       10.6      1.499         44        800: 100%|██████████| 99/99 [00:32<00:00,  3.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.53it/s]

                   all         26         80      0.532      0.498      0.454      0.263



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/50       2.4G     0.5022      10.39      1.488         31        800: 100%|██████████| 99/99 [00:32<00:00,  3.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.33it/s]

                   all         26         80       0.53      0.508      0.473      0.258



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/50       2.4G     0.5133      10.56      1.502         24        800: 100%|██████████| 99/99 [00:35<00:00,  2.79it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.53it/s]

                   all         26         80      0.534      0.493      0.482      0.257



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/50       2.4G     0.5008      10.31      1.482         60        800: 100%|██████████| 99/99 [00:32<00:00,  3.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.82it/s]

                   all         26         80      0.606      0.506      0.491      0.267



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/50       2.4G     0.5081      10.38      1.491         26        800: 100%|██████████| 99/99 [00:33<00:00,  2.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  5.00it/s]

                   all         26         80      0.482      0.511       0.45      0.262



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/50       2.4G     0.5093      10.55      1.506         25        800: 100%|██████████| 99/99 [00:32<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.77it/s]

                   all         26         80      0.573      0.567      0.492      0.283



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/50       2.4G     0.5058      10.47      1.491         74        800: 100%|██████████| 99/99 [00:33<00:00,  2.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.70it/s]

                   all         26         80      0.577      0.477      0.503      0.266



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/50       2.4G     0.5079      10.29      1.488         42        800: 100%|██████████| 99/99 [00:33<00:00,  2.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.76it/s]

                   all         26         80      0.578      0.483      0.469      0.262



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/50       2.4G     0.4999      10.23      1.482         22        800: 100%|██████████| 99/99 [00:34<00:00,  2.87it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.48it/s]

                   all         26         80       0.57      0.521      0.488      0.263



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/50       2.4G     0.4958      10.19      1.462         31        800: 100%|██████████| 99/99 [00:32<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.87it/s]

                   all         26         80      0.606      0.482      0.475      0.272



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/50       2.4G     0.4999      10.19      1.469         18        800: 100%|██████████| 99/99 [00:32<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.37it/s]

                   all         26         80      0.531      0.482       0.48      0.268



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/50       2.4G     0.4995      10.25      1.476         37        800: 100%|██████████| 99/99 [00:33<00:00,  2.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.01it/s]

                   all         26         80      0.567      0.494      0.482      0.272



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/50       2.4G     0.4996      10.11      1.473         53        800: 100%|██████████| 99/99 [00:32<00:00,  3.08it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.73it/s]

                   all         26         80      0.565      0.498      0.499      0.268



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/50       2.4G     0.4968      9.925      1.452         12        800: 100%|██████████| 99/99 [00:35<00:00,  2.78it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.80it/s]

                   all         26         80      0.487       0.57      0.505      0.263



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/50       2.4G      0.497      10.07      1.462         33        800: 100%|██████████| 99/99 [00:32<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.29it/s]

                   all         26         80      0.595      0.467       0.48      0.264



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/50       2.4G     0.4865       10.1      1.452         42        800: 100%|██████████| 99/99 [00:32<00:00,  3.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.92it/s]

                   all         26         80      0.623      0.467      0.497      0.275



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/50       2.4G     0.5095      10.43      1.497         44        800: 100%|██████████| 99/99 [00:33<00:00,  2.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.38it/s]

                   all         26         80      0.587       0.45      0.475       0.27



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/50       2.4G      0.502      10.19      1.468         66        800: 100%|██████████| 99/99 [00:33<00:00,  2.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.68it/s]

                   all         26         80      0.449      0.528      0.493      0.274



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/50       2.4G      0.507       10.3      1.494         19        800: 100%|██████████| 99/99 [00:33<00:00,  2.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  5.60it/s]

                   all         26         80       0.41       0.55      0.472      0.266



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/50       2.4G     0.5019      10.09      1.494         38        800: 100%|██████████| 99/99 [00:34<00:00,  2.89it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.79it/s]

                   all         26         80      0.572      0.456      0.461      0.255



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/50       2.4G     0.4979      9.984      1.458         27        800: 100%|██████████| 99/99 [00:33<00:00,  2.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.50it/s]

                   all         26         80      0.576      0.478      0.475      0.255



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/50       2.4G     0.4986      10.07      1.477         49        800: 100%|██████████| 99/99 [00:32<00:00,  3.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  3.61it/s]

                   all         26         80      0.643      0.501      0.498      0.257


Closing dataloader mosaic
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/50       2.4G     0.4051      7.838       1.25         26        800: 100%|██████████| 99/99 [00:28<00:00,  3.48it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  5.58it/s]


                   all         26         80      0.546      0.417      0.445      0.245

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/50       2.4G     0.4079      7.815      1.248         22        800: 100%|██████████| 99/99 [00:26<00:00,  3.67it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  4.37it/s]

                   all         26         80      0.607      0.508      0.484      0.245



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/50       2.4G     0.4007      7.634      1.233         36        800: 100%|██████████| 99/99 [00:26<00:00,  3.69it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  6.74it/s]

                   all         26         80      0.626      0.516      0.492      0.261



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/50       2.4G     0.4047      7.798      1.241         37        800: 100%|██████████| 99/99 [00:28<00:00,  3.48it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  8.07it/s]

                   all         26         80       0.45      0.523      0.447      0.254
EarlyStopping: Training stopped early as no improvement observed in last 20 epochs. Best results observed at epoch 24, best model saved as best.pt.
To update EarlyStopping(patience=20) pass a new patience value, i.e. `patience=300` or use `patience=0` to disable EarlyStopping.



44 epochs completed in 0.416 hours.
Optimizer stripped from runs/yolov11_custom_finetuned/weights/last.pt, 5.5MB
Optimizer stripped from runs/yolov11_custom_finetuned/weights/best.pt, 5.5MB

Validating runs/yolov11_custom_finetuned/weights/best.pt...
Ultralytics 8.3.89 🚀 Python-3.11.12 torch-2.6.0+cu124 CUDA:0 (Tesla T4, 15095MiB)
YOLO11n summary (fused): 100 layers, 2,582,932 parameters, 0 gradients, 6.3 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:01<00:00,  1.78it/s]


                   all         26         80      0.587      0.522      0.504      0.281
                  hole          3          5      0.514        0.8      0.679      0.422
                 crack         23         60      0.555      0.467       0.46       0.25
        block cracking          9         15      0.691        0.3      0.373      0.172
Speed: 0.3ms preprocess, 38.6ms inference, 0.0ms loss, 0.9ms postprocess per image
Results saved to runs/yolov11_custom_finetuned


# YOLOv11 Training Parameters
https://docs.ultralytics.com/modes/train/#train-settings
<br>
https://www.ultralytics.com/glossary/data-augmentation


# YOLOv11 Training Parameters with Practical Guidelines

| Argument | Type | Default | Description | Practical Guideline |
|----------|------|---------|-------------|---------------------|
| `model` | str | None | Specifies the model file for training. Accepts a path to either a .pt pretrained model or a .yaml configuration file. | For most tasks, use a pretrained YOLOv11 model (e.g., 'yolov11n.pt') for transfer learning rather than training from scratch. |
| `data` | str | None | Path to the dataset configuration file (e.g., coco8.yaml) containing dataset-specific parameters. | Create a custom YAML file with your dataset paths and class names; follow the COCO format for best compatibility. |
| `epochs` | int | 100 | Total number of training epochs, each representing a full pass over the dataset. | Use 50-100 for transfer learning; 100-300 for training from scratch; monitor validation metrics to avoid overfitting. |
| `time` | float | None | Maximum training time in hours, overriding the epochs argument. | Useful for time-boxed experiments; set to 24-48 hours for complex datasets when exact epoch count is less important than time constraints. |
| `patience` | int | 100 | Number of epochs to wait without improvement before early stopping. | Set to 20-50 for most tasks; lower (10-15) for quick experiments; higher (50-100) for complex datasets with slow convergence. |
| `batch` | int | 16 | Batch size for training (integer value, -1 for auto-sizing, or fraction for GPU utilization). | Use largest value that fits in memory; 8-16 for 8GB GPU; 32-64 for 24GB+ GPUs; smaller batches (4-8) can improve generalization. |
| `imgsz` | int or list | 640 | Target image size for training, affecting model accuracy and speed. | Use 640-1280 for general detection; larger (1280+) for small objects; smaller (416-512) for speed-critical applications. |
| `save` | bool | True | Enables saving of training checkpoints and final model weights. | Always keep True unless running purely experimental tests where saving is unnecessary. |
| `save_period` | int | -1 | Frequency of saving model checkpoints in epochs. | Set to 10-20 for long trainings to capture model evolution; lower (5) for critical/expensive trainings. |
| `cache` | bool | False | Enables caching of dataset images (True/ram, disk, or False). | Use 'disk' for large datasets; True/ram for small datasets (<5GB) with sufficient RAM; False for most typical cases. |
| `device` | int or str or list | None | Specifies computational device(s) for training. | For single GPU use '0'; multi-GPU use [0,1,2]; CPU-only use 'cpu'; use 'mps' for Apple Silicon. |
| `workers` | int | 8 | Number of worker threads for data loading per GPU. | Set to CPU core count minus 2; reduce to 0-4 if experiencing memory issues; increase to 12-16 for high-speed storage systems. |
| `project` | str | None | Name of the project directory for saving training outputs. | Use descriptive names for specific tasks ("traffic-detection", "crack-segmentation") to organize different research directions. |
| `name` | str | None | Name of the training run, creating a subdirectory within the project folder. | Use versioned names ("v1", "small-batch", "high-lr") that describe the specific training configuration for easy comparison. |
| `exist_ok` | bool | False | Allows overwriting of an existing project/name directory. | Set True during active development; False for production to prevent accidental overwriting of important results. |
| `pretrained` | bool or str | True | Determines whether to start training from a pretrained model. | Use True for most cases; False only when training on very specialized data unlike any pretraining dataset. |
| `optimizer` | str | 'auto' | Choice of optimizer for training. | Use 'SGD' for best generalization; 'Adam' for faster convergence; 'auto' works well for standard cases. |
| `seed` | int | 0 | Sets the random seed for training reproducibility. | Change only when comparing different architectures to ensure fair comparison with identical initialization. |
| `deterministic` | bool | True | Forces deterministic algorithm use, ensuring reproducibility. | Keep True for research and benchmarking; can set False for slight speed improvements in production training. |
| `single_cls` | bool | False | Treats all classes as a single class during training. | Set True only for presence/absence detection problems; False for all multi-class detection tasks. |
| `classes` | list[int] | None | Specifies class IDs to train on. | Use to focus on specific classes in a larger dataset; helpful for transfer learning on a subset of original classes. |
| `rect` | bool | False | Enables rectangular training, optimizing batch composition. | Set True for inference speed optimization; keep False during main training for better data augmentation. |
| `multi_scale` | bool | False | Enables multi-scale training by varying image size during training. | Enable for models that need robust multi-scale detection (traffic, aerial); unnecessary for fixed-distance applications. |
| `cos_lr` | bool | False | Utilizes a cosine learning rate scheduler. | Set True for most trainings; provides smoother LR decay and often better final convergence than step-based schedules. |
| `close_mosaic` | int | 10 | Disables mosaic augmentation in the last N epochs. | Set to 5-10% of total epochs; critical for stabilizing final training phase and improving precision. |
| `resume` | bool | False | Resumes training from the last saved checkpoint. | Use when training was interrupted; ensures consistent continuation of optimization process. |
| `amp` | bool | True | Enables Automatic Mixed Precision training. | Keep True on modern GPUs for memory savings and speed; set False only if experiencing numerical stability issues. |
| `fraction` | float | 1.0 | Fraction of the dataset to use for training. | Use 0.1-0.2 for quick experiments; 0.5 for prototype development; 1.0 for final training. |
| `profile` | bool | False | Enables ONNX and TensorRT speed profiling. | Enable only when planning to deploy with these frameworks to identify potential optimization bottlenecks. |
| `freeze` | int or list | None | Freezes the first N layers or specified layer indices. | Freeze 0-10 backbone layers when fine-tuning; more layers (10-15) for very similar datasets to the pretrained one. |
| `lr0` | float | 0.01 | Initial learning rate. | Use 0.01 for SGD with new models; 0.001-0.005 for fine-tuning; 0.0001-0.0005 for delicate final tuning. |
| `lrf` | float | 0.01 | Final learning rate as a fraction of initial rate. | Set 0.01-0.1 for typical training; lower (0.001-0.01) for longer trainings for smoother decay. |
| `momentum` | float | 0.937 | Momentum factor for SGD or beta1 for Adam. | Keep at 0.9-0.95 for most cases; lower (0.8-0.9) for noisy data; higher (0.95-0.99) for stable, well-behaved loss landscapes. |
| `weight_decay` | float | 0.0005 | L2 regularization factor, penalizing large weights. | Use 0.0005-0.001 for most cases; increase to 0.001-0.005 for small datasets prone to overfitting. |
| `warmup_epochs` | float | 3.0 | Number of epochs for learning rate warmup. | Set to 3-5 for stable training start; higher (5-10) for very large batch sizes or high learning rates. |
| `warmup_momentum` | float | 0.8 | Initial momentum for warmup phase. | Keep at default 0.8 for most cases; rarely needs adjustment. |
| `warmup_bias_lr` | float | 0.1 | Learning rate for bias parameters during warmup. | Keep at default 0.1; higher values (0.2-0.5) can sometimes improve initial detection of rare classes. |
| `box` | float | 7.5 | Weight of the box loss component in the loss function. | Use 5.0-7.5 for accurate localization; higher (8.0-10.0) when precise boundaries are critical; lower (3.0-5.0) to prioritize recall. |
| `cls` | float | 0.5 | Weight of the classification loss in the total loss function. | Increase to 1.0-2.0 to improve classification accuracy; decrease to 0.3-0.5 when localization is more important than classification. |
| `dfl` | float | 1.5 | Weight of the distribution focal loss. | Keep at default 1.5 for most cases; increase to 2.0-3.0 for highly overlapping objects requiring precise distribution. |
| `pose` | float | 12.0 | Weight of the pose loss in models trained for pose estimation. | Only relevant for pose models; increase to 15.0-20.0 for applications requiring very precise keypoint localization. |
| `kobj` | float | 2.0 | Weight of the keypoint objectness loss in pose estimation. | Only for pose models; balance with pose parameter (typically 1:5 to 1:8 ratio) depending on application needs. |
| `nbs` | int | 64 | Nominal batch size for normalization of loss. | Rarely needs changing; adjust only if experiencing unstable gradients with very large or small batch sizes. |
| `overlap_mask` | bool | True | Determines mask merging strategy for overlapping objects. | Keep True for most segmentation tasks; set False only for specialized cases with transparent overlapping objects. |
| `mask_ratio` | int | 4 | Downsample ratio for segmentation masks. | Use 4 for balance of speed and accuracy; lower (2) for fine detail preservation; higher (8) for speed on large masks. |
| `dropout` | float | 0.0 | Dropout rate for regularization in classification. | Set to 0.1-0.3 for classifiers prone to overfitting; keep at 0.0 for most detection tasks unless overfitting is observed. |
| `val` | bool | True | Enables validation during training. | Always keep True to monitor training progress; False only in rare cases where validation would be redundant. |
| `plots` | bool | False | Generates and saves plots of training metrics and examples. | Enable for visualizing training progress and model performance; especially useful for debugging or tuning hyperparameters. |

# YOLO Augmentation Parameters with Practical Guidelines

| Argument | Type | Default | Range | Description | Practical Guideline |
|----------|------|---------|-------|-------------|---------------------|
| `hsv_h` | float | 0.015 | 0.0 - 1.0 | Adjusts the hue of the image by a fraction of the color wheel, introducing color variability. Helps the model generalize across different lighting conditions. | Keep values small (0.015-0.05) for realistic colors; use higher values (0.1-0.2) only when color variation is critical for detection. |
| `hsv_s` | float | 0.7 | 0.0 - 1.0 | Alters the saturation of the image by a fraction, affecting the intensity of colors. Useful for simulating different environmental conditions. | Use higher values (0.5-0.7) for outdoor datasets with varying lighting conditions; lower (0.2-0.4) for controlled indoor environments. |
| `hsv_v` | float | 0.4 | 0.0 - 1.0 | Modifies the value (brightness) of the image by a fraction, helping the model to perform well under various lighting conditions. | Set to 0.3-0.5 for datasets with mixed lighting; increase to 0.6-0.8 for night/low-light detection scenarios. |
| `degrees` | float | 0.0 | 0.0 - 180 | Rotates the image randomly within the specified degree range, improving the model's ability to recognize objects at various orientations. | Use 0-10° for objects with directional properties (road markings, text); 10-30° for orientation-independent objects (people, animals); avoid for horizon-aligned scenes. |
| `translate` | float | 0.1 | 0.0 - 1.0 | Translates the image horizontally and vertically by a fraction of the image size, aiding in learning to detect partially visible objects. | Set to 0.1-0.2 for centered objects; increase to 0.3-0.5 for objects that appear at image edges or are partially visible. |
| `scale` | float | 0.5 | >=0.0 | Scales the image by a gain factor, simulating objects at different distances from the camera. | Use 0.4-0.6 for general detection tasks; higher (0.7-0.9) when object size varies significantly in your dataset. |
| `shear` | float | 0.0 | -180 - +180 | Shears the image by a specified degree, mimicking the effect of objects being viewed from different angles. | Keep low (1.0-3.0) for most tasks; increase to 5.0-10.0 for datasets with varied camera angles; use cautiously as it can distort object shapes. |
| `perspective` | float | 0.0 | 0.0 - 0.001 | Applies a random perspective transformation to the image, enhancing the model's ability to understand objects in 3D space. | Use very small values (0.0001-0.0005) to maintain realistic object appearances; higher values distort objects excessively. |
| `flipud` | float | 0.0 | 0.0 - 1.0 | Flips the image upside down with the specified probability, increasing the data variability without affecting the object's characteristics. | Generally keep at 0.0; only use (0.3-0.5) for orientation-invariant objects like aerial imagery or symmetric objects. |
| `fliplr` | float | 0.5 | 0.0 - 1.0 | Flips the image left to right with the specified probability, useful for learning symmetrical objects and increasing dataset diversity. | Set to 0.5 for most datasets; reduce to 0.0 for directional objects (text, arrows); essential for doubling effective dataset size. |
| `bgr` | float | 0.0 | 0.0 - 1.0 | Flips the image channels from RGB to BGR with the specified probability, useful for increasing robustness to incorrect channel ordering. | Usually keep at 0.0; set to 0.1-0.3 only when developing models that might be deployed across different image processing pipelines. |
| `mosaic` | float | 1.0 | 0.0 - 1.0 | Combines four training images into one, simulating different scene compositions and object interactions. | Set to 0.7-0.9 for small object detection tasks; reduce to 0.5-0.6 for datasets with thin or fine features (cracks, wires); disable (0.0) in the final few epochs. |
| `mixup` | float | 0.0 | 0.0 - 1.0 | Blends two images and their labels, creating a composite image. Enhances the model's ability to generalize by introducing label noise and visual variability. | Use 0.1-0.2 for small datasets; 0.3-0.4 for complex scenes; avoid or keep very low (0.05) for fine-detail detection tasks. |
| `cutmix` | float | 0.0 | 0.0 - 1.0 | Combines portions of two images, creating a partial blend while maintaining distinct regions. | Set to 0.2-0.3 for classification tasks; lower (0.1) for detection of small objects; higher (0.4-0.5) for robust multi-class detection. |
| `copy_paste` | float | 0.0 | 0.0 - 1.0 | *Segmentation only*. Copies and pastes objects across images to increase object instances. | Use 0.2-0.4 for rare classes with few samples; increase to 0.5-0.7 for extreme class imbalance; critical for improving recall of uncommon objects. |
| `copy_paste_mode` | str | flip | - | *Segmentation only*. Specifies the `copy-paste` strategy to use. Options include `'flip'` and `'mixup'`. | Use 'flip' for most segmentation tasks; try 'mixup' when diversity in object appearance is more important than position. |
| `auto_augment` | str | randaugment | - | *Classification only*. Applies a predefined augmentation policy to enhance model performance through visual diversity. | 'randaugment' works well for general tasks; try 'autoaugment' for known datasets (CIFAR, ImageNet); 'augmix' for highest robustness to distribution shifts. |
| `erasing` | float | 0.4 | 0.0 - 0.9 | *Classification only*. Randomly erases regions of the image during training to encourage the model to focus on less obvious features. | Set to 0.3-0.5 for general classification; increase to 0.6-0.8 for objects with distinctive parts that should not be relied on exclusively. |

## Recommended Parameters

### For General Object Detection
- `mosaic=0.8` - Strong but not overwhelming mosaic augmentation
- `mixup=0.1` - Light blending to maintain feature clarity
- `degrees=10.0` - Moderate rotation for orientation robustness
- `fliplr=0.5` - Standard horizontal flipping
- `translate=0.1` - Slight position variation
- `scale=0.5` - Balanced size variation
- `pretrained=True` - Use transfer learning from COCO dataset
- `batch=16` (or highest that fits on GPU)
- `imgsz=640`
- `epochs=100`
- `patience=50`
- `lr0=0.002`, `lrf=0.01`
- `cos_lr=True`
- `close_mosaic=10`

### For Small Object Detection (e.g., defects, cracks)
- `mosaic=0.6` - Lower mosaic to prevent tiny objects from disappearing
- `mixup=0.05` - Very light blending to preserve small features
- `degrees=5.0` - Minimal rotation
- `scale=0.6` - Emphasize varying sizes
- `hsv_v=0.5` - Enhanced brightness variation
- `copy_paste=0.3` - Increase instances of rare defects
- `imgsz=1280` (larger input size)
- `batch=8` (accommodate larger images)
- `box=10.0` (emphasize localization accuracy)
- `iou=0.35` (more lenient matching criteria)
- `epochs=150` (longer training)
- `close_mosaic=15` (more stabilization epochs)

### For Robust Real-World Deployment
- `hsv_h=0.02, hsv_s=0.7, hsv_v=0.4` - Comprehensive color variation
- `mosaic=0.7` - Balanced mosaic augmentation
- `mixup=0.2` - Moderate blending for robustness
- `translate=0.2` - Increased position variation
- `perspective=0.0003` - Slight perspective change
- `degrees=15.0` - Stronger rotation for viewpoint invariance
- Final training phase after main training:
  - `lr0=0.0001` (very low learning rate)
  - `epochs=20` (short fine-tuning)
  - `rect=True` (rectangular inference optimization)
  - `multi_scale=True` (robust to different image sizes)
  - `batch=4` (smallest stable batch size for best generalization)

### For Limited Dataset Training
- `mosaic=0.9` - Heavy mosaic to maximize data combination
- `mixup=0.2` - Moderate blending for artificial diversity
- `cutmix=0.2` - Additional variation technique
- `copy_paste=0.4` - Replicate scarce objects
- `fliplr=0.5, scale=0.7` - Maximize effective data variation